### PR TITLE
ref(workflow_engine): Updates to the data models

### PR DESCRIPTION
**Pull request by saponifi3d**
## Description
As we continue to spec out the migrations for issue alerts / metric alerts, we found a few more data model changes.

### CODE REVIEW

File: src/sentry/workflow_engine/migrations/0013_aci_model_updates_from_milestones.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                    ALTER TABLE "workflow_engine_detector" DROP COLUMN "enabled";
+                    """,
+                    hints={"tables": ["workflow_engine_detector"]},
+                ),
+            ],
+            state_operations=[
+                migrations.AddField(
+                    model_name="detector",
+                    name="enabled",
+                    field=sentry.db.models.fields.bounded.BoundedPositiveIntegerField(default=1),
```
Comment (wedamija) [inline src/sentry/workflow_engine/migrations/0013_aci_model_updates_from_milestones.py:59]: This should be default=True and a boolean field

File: src/sentry/workflow_engine/models/detector.py [hunk 2]
```diff
@@ -21,17 +21,29 @@
 
 @region_silo_model
 class Detector(DefaultFieldsModel, OwnerModel):
-    __relocation_scope__ = RelocationScope.Organization
+    __relocation_scope__ = RelocationScope.Excluded
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:24]: Why'd we move to excluded instead of project?

File: src/sentry/workflow_engine/migrations/0013_aci_model_updates_from_milestones.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                    ALTER TABLE "workflow_engine_workflow" DROP COLUMN "enabled";
+                    """,
+                    hints={"tables": ["workflow_engine_workflow"]},
+                ),
+            ],
+            state_operations=[
+                migrations.AddField(
+                    model_name="workflow",
+                    name="enabled",
+                    field=sentry.db.models.fields.bounded.BoundedPositiveIntegerField(default=1),
```
Comment (wedamija) [inline src/sentry/workflow_engine/migrations/0013_aci_model_updates_from_milestones.py:93]: This should be True and a boolean field

File: src/sentry/workflow_engine/models/detector.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/workflow_engine/models/detector.py:24]: Hmm, I think we're just meant to use `Organization` actually like https://github.com/getsentry/sentry/blob/1c0cb149b011907a16882411a46c9b5181ca268e/src/sentry/models/options/project_option.py#L166-L175

File: src/sentry/workflow_engine/models/detector.py [hunk 4]
```diff
@@ -23,15 +23,27 @@
...
 
     # The data sources that the detector is watching
     data_sources = models.ManyToManyField(
         "workflow_engine.DataSource", through="workflow_engine.DataSourceDetector"
     )
 
-    # The conditions that must be met for the detector to be considered 'active'
+    # If the detector is not enabled, it will not be evaluated. This is how we "snooze" a detector
+    enabled = models.BooleanField(default=True)
+
```
Comment (ceorourke) [inline src/sentry/workflow_engine/models/detector.py:39]: Weren't we going to add a `created_by` field to this model to preserve the historical data?

File: src/sentry/workflow_engine/models/detector.py [hunk 5]
```diff
@@ -23,15 +23,27 @@
...
+    # Optionally set a description of the detector, this will be used in notifications
+    description = models.TextField(blank=True, null=True)
+
+    # This represents a time delta, in seconds. If not null, this is used to determine which time
+    # window to query to compare the result from the current time_window to.
+    comparison_delta = models.IntegerField(null=True)
```
[the remark this answers is not in the document]
Comment (saponifi3d) [reply src/sentry/workflow_engine/models/detector.py:45]: Yeah, i agree this is a bit specific, i couldn't think of a better place for it to live at the time...

what do you think of including this on the data_source?

File: src/sentry/workflow_engine/models/detector.py (hunk 4 above)
Comment (saponifi3d) [reply src/sentry/workflow_engine/models/detector.py:39]: 👍 i couldn't remember if it was this or just the workflow that needed it. will update.

File: src/sentry/workflow_engine/models/detector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/workflow_engine/models/detector.py:45]: I think that only makes sense if the data source is going to be performing the comparison comparation, otherwise it doesn't care. I'm not really sure what the best pattern is here, but it seems like we need somewhere to put fields that are specific to a given detector type

[the remark this answers is not in the document]
Comment (saponifi3d) [reply src/sentry/workflow_engine/models/detector.py:45]: chatted in slack about this; but wanted to close the loop...

it seems like there are really two solutions - creating a separate table for metric detectors and base detectors, or creating a json field with a schema. The former makes querying / data lookup a bit difficult, and we've seen a few of the downsides to that in the alerts api today.

Going to focus on making a `config` field that's a json blob. This field can have a schema that is defined on each sub-class to ensure consistence there, and enforce the type checking on the base `detector` model.

File: src/sentry/workflow_engine/models/detector.py (hunk 4 above)
Comment (saponifi3d) [reply src/sentry/workflow_engine/models/detector.py:43]: added the `created_by` -- for the `threshold_period` we'll use the `config` field, and a CONFIG_SCHEMA to define the attributes there. this will need to be part of the `MetricDetector`

File: src/sentry/workflow_engine/models/detector.py [hunk 6]
```diff
@@ -4,34 +4,46 @@
...
 
 
 @region_silo_model
-class Detector(DefaultFieldsModel, OwnerModel):
+class Detector(DefaultFieldsModel, OwnerModel, JSONConfigBase):
     __relocation_scope__ = RelocationScope.Organization
 
-    organization = FlexibleForeignKey("sentry.Organization")
+    # TODO - Finish removing this field
+    organization = FlexibleForeignKey("sentry.Organization", on_delete=models.CASCADE, null=True)
```
[the remark this answers is not in the document]
Comment (saponifi3d) [reply src/sentry/workflow_engine/models/detector.py:31]: because we added the `project` - there are a few interactions that make org + project a little tricky, and we can reference the org via the project.

**Review by ceorourke — APPROVED** [review]
lgtm in terms of changes I need for metric alerts - feel free to wait until Dan approves too

File: src/sentry/workflow_engine/models/json_config.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+from abc import abstractproperty
+from typing import Any
+
+from django.db import models
+from jsonschema import ValidationError, validate
+
+
+class JSONConfigBase(models.Model):
```
Comment (saponifi3d) [inline src/sentry/workflow_engine/models/json_config.py:8]: I had added tests for this here: https://github.com/getsentry/sentry/pull/80710/commits/8abfecc8f696c3496615e2ec4d5e8eb5b9e00991 - but removed them cause it was causing a bunch of other failures. any recommendations on how to fix these?

File: src/sentry/workflow_engine/migrations/0014_aci_model_updates_from_milestones.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                    ALTER TABLE "workflow_engine_detector" DROP COLUMN "config";
+                    """,
+                    hints={"tables": ["workflow_engine_detector"]},
+                ),
+            ],
+            state_operations=[
+                migrations.AddField(
+                    model_name="detector",
+                    name="config",
+                    field=models.JSONField(blank=True, default=dict, null=True),
```
Comment (wedamija) [inline src/sentry/workflow_engine/migrations/0014_aci_model_updates_from_milestones.py:48]: Why does this have `null=True` and the sql has `NOT NULL`?

I also updated our migration framework, so you can actually skip this whole sql step: https://develop.sentry.dev/api-server/application-domains/database-migrations/#adding-columns-with-a-default

tl;dr: `models.JSONField(db_default=dict)`

File: src/sentry/workflow_engine/models/json_config.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    config = models.JSONField(default=dict, blank=True, null=True)
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/json_config.py:9]: You can define this as `models.JSONField(db_default=dict)`. No need for it to be nullable

File: src/sentry/workflow_engine/models/detector.py [hunk 10]
```diff
@@ -4,34 +4,46 @@
...
     data_sources = models.ManyToManyField(
         "workflow_engine.DataSource", through="workflow_engine.DataSourceDetector"
     )
 
-    # The conditions that must be met for the detector to be considered 'active'
+    # If the detector is not enabled, it will not be evaluated. This is how we "snooze" a detector
+    enabled = models.BooleanField(default=True)
+
+    # Optionally set a description of the detector, this will be used in notifications
+    description = models.TextField(blank=True, null=True)
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:45]: Blank means empty string, null means can be null. Do you need both of these set?

File: src/sentry/workflow_engine/migrations/0014_aci_model_updates_from_milestones.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+                    hints={"tables": ["workflow_engine_detector"]},
+                ),
+            ],
+            state_operations=[
+                migrations.AddField(
+                    model_name="detector",
+                    name="enabled",
+                    field=models.BooleanField(default=True),
+                ),
+            ],
```
Comment (wedamija) [inline src/sentry/workflow_engine/migrations/0014_aci_model_updates_from_milestones.py:82]: Same comment from before - you can define this with `db_default=True` and not have the sql

File: src/sentry/workflow_engine/models/detector.py [hunk 12]
```diff
@@ -4,34 +4,46 @@
...
     name = models.CharField(max_length=200)
 
     # The data sources that the detector is watching
     data_sources = models.ManyToManyField(
         "workflow_engine.DataSource", through="workflow_engine.DataSourceDetector"
     )
 
-    # The conditions that must be met for the detector to be considered 'active'
+    # If the detector is not enabled, it will not be evaluated. This is how we "snooze" a detector
+    enabled = models.BooleanField(default=True)
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:42]: `db_default=True` instead

File: src/sentry/workflow_engine/models/action.py [hunk 13]
```diff
@@ -24,17 +24,13 @@ class Type(models.TextChoices):
...
-    These actions will be used internally, to trigger other aspects of the system.
-    For example, creating a new issue in the Issue Platform or a detector emitting an event.
-    """
-    required = models.BooleanField(default=False)
-
     # The type field is used to denote the type of action we want to trigger
     type = models.TextField(choices=Type.choices)
     data = models.JSONField(default=dict)
 
+    # TODO - finish removing this field
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/action.py:31]: JFYI: I've got an operation coming soon that will make deleting these things simpler https://github.com/getsentry/sentry/pull/81098

File: src/sentry/workflow_engine/models/workflow.py [hunk 14]
```diff
@@ -17,8 +22,23 @@ class Workflow(DefaultFieldsModel):
     name = models.CharField(max_length=200)
     organization = FlexibleForeignKey("sentry.Organization")
 
+    # If the workflow is not enabled, it will not be evaluated / invoke actions. This is how we "snooze" a workflow
+    enabled = models.BooleanField(default=True)
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:26]: `db_default=True`

File: src/sentry/workflow_engine/models/workflow.py [hunk 15]
```diff
@@ -17,8 +22,23 @@ class Workflow(DefaultFieldsModel):
...
+
     # Required as the 'when' condition for the workflow, this evalutes states emitted from the detectors
-    when_condition_group = FlexibleForeignKey(DataConditionGroup, blank=True, null=True)
+    when_condition_group = FlexibleForeignKey(
+        "workflow_engine.DataConditionGroup", blank=True, null=True
+    )
+
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:32]: No need for `blank=True`

File: src/sentry/workflow_engine/models/workflow.py [hunk 16]
```diff
@@ -17,8 +22,23 @@ class Workflow(DefaultFieldsModel):
...
+    environment = FlexibleForeignKey(
+        "sentry.Environment", db_constraint=False, blank=True, null=True
+    )
+
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:36]: No need for `blank=True`. Why `db_constraint=False`?

File: src/sentry/workflow_engine/models/workflow.py [hunk 17]
```diff
@@ -17,8 +22,23 @@ class Workflow(DefaultFieldsModel):
...
+    created_by = HybridCloudForeignKey(settings.AUTH_USER_MODEL, null=True, on_delete="SET_NULL")
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:37]: This should be `created_by_id` since this uses a `HybridCloudForeignKey`

File: src/sentry/workflow_engine/models/detector.py [hunk 18]
```diff
@@ -40,8 +52,18 @@ class Detector(DefaultFieldsModel, OwnerModel):
         unique=True,
         on_delete=models.SET_NULL,
     )
+
+    # The type of detector that is being used, this is used to determine the class
+    # to load for the detector
     type = models.CharField(max_length=200)
 
+    # The user that created the detector
+    created_by = HybridCloudForeignKey(settings.AUTH_USER_MODEL, null=True, on_delete="SET_NULL")
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:61]: `created_by_id`

**Review by wedamija — APPROVED** [review]

File: src/sentry/workflow_engine/models/workflow.py [hunk 19]
```diff
@@ -17,8 +24,17 @@ class Workflow(DefaultFieldsModel):
...
+
     # Required as the 'when' condition for the workflow, this evalutes states emitted from the detectors
-    when_condition_group = FlexibleForeignKey(DataConditionGroup, blank=True, null=True)
+    when_condition_group = FlexibleForeignKey(
+        "workflow_engine.DataConditionGroup", blank=True, null=True
+    )
+
+    environment_id = BoundedPositiveIntegerField(null=True)
+    created_by = HybridCloudForeignKey(settings.AUTH_USER_MODEL, null=True, on_delete="SET_NULL")
+    frequency = BoundedPositiveIntegerField(default=30)  # in minutes
```
[the remark this answers is not in the document]
Comment (cathteng) [reply src/sentry/workflow_engine/models/workflow.py]: i supposed we could add this to the config and only apply it if it exists for issue alerts. i can't see this being useful for anything else
